pkg/blockstorage/zone/zone.go

```diff
@@ -39,153 +36,89 @@ type (
 )
 
 // FromSourceRegionZone gets the zones from the given region and sourceZones
-// It will return a minimum of 0 and a maximum of zones equal to the length of sourceZones.
+// It will return a minimum of 0 and a maximum of zones equal to the length zones available from kubernetes.
 // Depending on the length of the slice returned, the blockstorage providers will decide if
 // a regional volume or a zonal volume should be created.
-func FromSourceRegionZone(ctx context.Context, m Mapper, region string, sourceZones ...string) ([]string, error) {
+func FromSourceRegionZone(ctx context.Context, m Mapper, kubeCli kubernetes.Interface, sourceRegion string, sourceZones ...string) ([]string, error) {
 	newZones := make(map[string]struct{})
-	cli, err := kube.NewClient()
-	if err == nil {
-		nzs, rs, errzr := NodeZonesAndRegion(ctx, cli)
-		if errzr != nil {
-			log.WithError(errzr).Print("Ignoring error getting Node availability zones.")
-		}
-		if len(nzs) != 0 {
-			for _, zone := range sourceZones {
-				var z string
-				// getZoneFromKnownNodeZones() func makes sure that all zones
-				// added to newZones[] are unique and not repeated.
-				z, err = getZoneFromKnownNodeZones(ctx, zone, nzs, newZones)
-				if err == nil && isZoneValid(ctx, m, z, rs) {
-					newZones[z] = struct{}{}
-				}
-				if err != nil {
-					log.WithError(err).Print("Ignoring error getting Zone from KnownNodeZones.")
-				}
-			}
-		}
+	validZoneNames, err := m.FromRegion(ctx, sourceRegion)
+	if err != nil || len(validZoneNames) == 0 {
+		return nil, errors.Wrapf(err, "No provider zones for region (%s)", sourceRegion)
 	}
+	if !isNil(kubeCli) {
+		getAvailableZones(ctx, newZones, kubeCli, validZoneNames, sourceZones, sourceRegion)
```

Comment:
This modifies `newZones`, right? I would just return it, otherwise this takes extra effort to read.

I would also push down handling of a nil kubeCli and return an error if something goes wrong.

Reply:
So this is taking into account that if kubecli is nil we would use the zones from the provider to compare the source zone against. If you'd rather we fail out at this point thats fine too